### ISSUE TITLE
How to test on new images?

### ISSUE THREAD

**Issue by pultarmi** [lixx2938/CGIntrinsics#9]
Hi,
thank you for your work, I like the results. We are thinking about utilizing the model to improve the registration pipeline for pairs of images. However, I am not successful in using the code for testing on our dataset. I would find it very useful if you could provide some info about how to use it on inputs which do not have the IIW/SAW dataset format - e.g. on a single image.

Best regards,
Milan Pultar

**Comment by HawaiiZeng** [comment]
> Hi,
> thank you for your work, I like the results. We are thinking about utilizing the model to improve the registration pipeline for pairs of images. However, I am not successful in using the code for testing on our dataset. I would find it very useful if you could provide some info about how to get predictions from inputs which are not in the IIW/SAW dataset format - e.g. how to get an output from a single image.
> 
> Best regards,
> Milan Pultar

Did you manage to make it work on single image? Can you share any script for others to use?

**Comment by pultarmi** [comment]
No, sorry. I haven't paid much attention to this since I wrote this question. But it looks like it should not be too hard to create such script.

**Comment by SystemErrorWang** [comment]
I succeeded in testing on a single image with the provided pre-trained model, but found severe checkerboard artifacts. It may be because of the image pre-process.

**Comment by anandbhattad** [comment]
I can load pretrained model. However, the outputs are two gray-scale (single) channel images. One looks like shading and another I am not sure (residual?). Is this correct? 
Minimal code to import pretrained model: 

`from CGIntrinsics.models.intrinsic_model import *
from CGIntrinsics.models.networks import *
from CGIntrinsics.models.base_model import *
cgnet = define_G(input_nc=3, output_nc=3, ngf=64, which_model_netG="unet_256")
cgnet.load_state_dict(torch.load("pretrained_models/cgintrinsics_iiw_saw_final_net_G.pth"))
cgnet = cgnet.cuda();
cgnet.eval();`

**Comment by jundanl** [comment]
> I succeeded in testing on a single image with the provided pre-trained model, but found severe checkerboard artifacts. It may be because of the image pre-process.

I believe the checkerboard artifacts are caused by the deconv layers and the unreasonable implementation of computing image gradient.

I removed all the checkerboard artifacts by addressing the above two problems.

**Comment by MayuOshima** [comment]
> > I succeeded in testing on a single image with the provided pre-trained model, but found severe checkerboard artifacts. It may be because of the image pre-process.
> 
> I personally think the checkerboard artifacts are caused by the deconv layers and the unreasonable implementation of computing image gradient.
> 
> I removed all the checkerboard artifacts by addressing the above two problems.

Hello, could you please share how you address the two problems ? Thank you!

**Comment by tlsshh** [comment]
> > > I succeeded in testing on a single image with the provided pre-trained model, but found severe checkerboard artifacts. It may be because of the image pre-process.
> > 
> > 
> > I personally think the checkerboard artifacts are caused by the deconv layers and the unreasonable implementation of computing image gradient.
> > I removed all the checkerboard artifacts by addressing the above two problems.
> 
> Hello, could you please share how you address the two problems ? Thank you!

1. Replace deconv layers with upsample. This can reduce artifacts but lead to larger computation.
2. It's hard to explain why the implementation of computing image gradients is unreasonable. Please compute the image gradient by `x_(i+1) - x_i`, rather that 'x_(i+2) - x_i'. Skipping the pixel `x_(i+1)` leads to artifacts.

This is my [NIID-Net](https://github.com/zju3dv/NIID-Net). Hope the modified [loss functions ](https://github.com/zju3dv/NIID-Net/blob/master/loss/criteria_intrinsic.py) can help you.
